### PR TITLE
Add paramInfoRepeated and use it in data file interface

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #21139.

### CODE REVIEW

**Review by loganharbour — APPROVED** [review]

**Review by loganharbour — DISMISSED** [review]

**Comment by dschwen** [comment]
The side effect is that in material classes this will always result in triple output.

**Comment by YaqiWang** [comment]
> The side effect is that in material classes this will always result in triple output.

Dang. I did not think of this. We can override the function in `MaterialBase` to skip issuing info for the other two copies of materials or we can let `DataFileInterface` check if `_parent` is a material, if so, issue the info only for the volumetric one. Which one you prefer?

File: framework/include/utils/MooseUtils.h [hunk 1]
```diff
@@ -242,6 +233,54 @@ std::string stripExtension(const std::string & s);
...
+  }
+  else if (rest.size() > 0)
+  {
+    if (rest[0] == '/')
+      mooseError("Cannot join an absolute path");
+    path = path + "/" + rest;
+  }
+
+  return path;
+}
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:282]: this is from your other PR. I think it'd be best to keep those issues separated.

File: framework/src/base/Registry.C [hunk 2]
```diff
@@ -110,17 +110,23 @@ Registry::addKnownLabel(const std::string & label)
 }
 
 void
-Registry::addDataFilePath(const std::string & fullpath)
+Registry::addDataFilePathFromApp(const std::string & fullpath)
```
Comment (dschwen) [inline framework/src/base/Registry.C:113]: This is unrelated to the linked issue and I strongly recommend against it as it messes up the design,m which is explicitly based on an automatically (if present) installed `data` dir.

File: framework/src/interfaces/DataFileInterface.C [hunk 3]
```diff
@@ -40,6 +38,17 @@ std::string
...
+  bool issue_info = true;
+
+  // we do not want to issue the information for the face and neighbor materials
+  // because the same information has been outputted by the element material
+  auto mat_ptr = dynamic_cast<const Material *>(&_parent);
+  if (mat_ptr)
+  {
+    if (mat_ptr->isBoundaryMaterial() && mat_ptr->blockRestricted())
+      issue_info = false;
+  }
```
Comment (dschwen) [inline framework/src/interfaces/DataFileInterface.C:50]: this should go in the interface constructor and set a member variable.

**Review by dschwen — CHANGES_REQUESTED** [review]
see comments

File: framework/src/interfaces/DataFileInterface.C (hunk 3 above)
Comment (YaqiWang) [reply framework/src/interfaces/DataFileInterface.C:50]: Yes.

File: framework/src/base/Registry.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/base/Registry.C:113]: I disagree. You called this function anyway. Maybe you or me misunderstood something?

File: framework/include/utils/MooseUtils.h (hunk 1 above)
Comment (YaqiWang) [reply framework/include/utils/MooseUtils.h:282]: You mean from the other issue? I do not need this commit and can drop it. But as I commented in the issue, if your PR cannot be merged soon, I rather this gets merged first. The screen printout is just pure stupid.

Comment (YaqiWang) [reply framework/include/utils/MooseUtils.h]: Oh, saw your comment on the issue. So we agree to have a quick fix. I think unless we need more unit tests on this function, we can ship it with this. Our test suites can make sure no mistakes in this function.

File: framework/src/base/Registry.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/base/Registry.C:113]: This function may not work with binary installation. Maybe I missed something because I cannot find codes for auto installed data dir.

**Comment by YaqiWang** [comment]
This should be ready again. I chatted with @permcody a little bit. We have doubts on whether the data folder registration works with binary distribution. @dschwen The data folder registration needs to be different for an executable in installation folder and an executable compiled from source. The function I added in `Registry` can be a step toward that. I do not have any fundamental change in this PR, keep in mind.

**Comment by YaqiWang** [comment]
Can I get a review or an approval? I've addressed the comments.

**Comment by YaqiWang** [comment]
@dschwen 
167941137-a8b32c14-37d0-46c5-9a09-3e4ad500ebf1

**Comment by dschwen** [comment]
> I think it is better to use `INSTALLABLE_DIRS` and avoid auto copy.

I don't think so, because the data file name relies on that known `data` name.

File: framework/include/base/Registry.h [hunk 4]
```diff
@@ -192,6 +192,9 @@ class Registry
   /// addKnownLabel whitelists a label as valid for purposes of the checkLabels function.
   static char addKnownLabel(const std::string & label);
 
+  /// register search paths for built-in data files based on the app file name
+  static void addDataFilePathFromApp(const std::string & path);
+
```
Comment (dschwen) [inline framework/include/base/Registry.h:197]: this has nothing to do with `paramInfoRepeated`

File: framework/include/utils/MooseUtils.h [hunk 5]
```diff
@@ -242,6 +233,59 @@ std::string stripExtension(const std::string & s);
...
+    }
+    else
+      break;
+
+  if (rest.size() > 0)
+    path = path + "/" + rest;
+
+  return path;
+}
+
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:288]: this has nothing to do with `paramInfoRepeated`

File: framework/src/base/Registry.C (hunk 2 above)
Comment (dschwen) [reply framework/src/base/Registry.C:113]: No, you don't need to call `addDataFilePath` directly. The design does not require it. What you need can be done without these changes.

**Review by dschwen — CHANGES_REQUESTED** [review]

**Comment by dschwen** [comment]
Here is my high level assessment: We have a way to do a task by doing `A`. You don't quite like `A` and would rather do it way `B`. But `B` is worse and ignores key design ideas that went into deciding to make way `A` the way to go. You add code that makes everything more verbose and confusing, because now users have two ways of doing the same thing. 👎

**Comment by YaqiWang** [comment]
OK, I will create a new PR so that we can continue our conversation there. This PR will be resolving that param info issue.

**Comment by YaqiWang** [comment]
Ok, I stripped out the other three commits and created an issue (https://github.com/idaholab/moose/issues/21187). I will submit a draft PR for that issue.

**Comment by YaqiWang** [comment]
> No, you don't need to call addDataFilePath directly. The design does not require it. What you need can be done without these changes.

No because you had hard-coded `../../data` there.

**Comment by YaqiWang** [comment]
@dschwen and @loganharbour this should be less controversial and ready.

**Comment by YaqiWang** [comment]
@loganharbour this should be ready for merge. The conversation we had with Daniel should be in another PR. The code changes here made that irrelevant.
